packages/react-components/src/components/xy-output-component.tsx

```diff
@@ -426,24 +432,108 @@ export class XYOutputComponent extends AbstractTreeOutputComponent<AbstractOutpu
         }
     }
 
+    private tooltip(x: number, y: number): void {
+        const xPos = this.positionXMove;
+        const timeForX = this.getTimeForX(xPos);
+        let timeLabel: string | undefined = timeForX.toString();
+        if (this.props.unitController.numberTranslator) {
+            timeLabel = this.props.unitController.numberTranslator(timeForX);
+        }
+        const chartWidth = this.lineChartRef.current.chartInstance.width;
+        const arraySize = this.state.xyData.labels.length;
+        const index = Math.max(Math.round((xPos / chartWidth) * (arraySize - 1)), 0);
+        const points: any = [];
+        let zeros = 0;
+
+        this.state.xyData.datasets.forEach((d: any) => {
+            // In case there are less data points than pixels in the chart,
+            // calculate nearest value.
+            const yValue = d.data[index];
+            const rounded = isNaN(yValue) ? 0 : (Math.round(Number(yValue) * 100) / 100);
+            // If there are less than 10 lines in the chart, show all values, even if they are equal to 0.
+            // If there are more than 10 lines in the chart, summarise the ones that are equal to 0.
+            if (this.state.xyData.datasets.length <= 10 || rounded > 0) {
+                const point: any = {
+                    label: d.label,
+                    color: d.borderColor,
+                    background: d.backgroundColor,
+                    value: rounded.toString(),
+                };
+                points.push(point);
+            }
+            else {
+                zeros += 1;
+            }
+        });
+        // Sort in decreasing order
+        points.sort((a: any, b: any) => Number(b.value) - Number(a.value));
+
+        // Adjust tooltip position if mouse is too close to the bottom of the window
+        let topPos = undefined;
+        let bottomPos = undefined;
+        if (y > window.innerHeight - 350) {
+            bottomPos = window.innerHeight - y;
+        }
+        else {
+            topPos = window.pageYOffset + y - 40;
+        }
+
+        // Adjust tooltip position if mouse is too close to the left edge of the chart
+        let leftPos = undefined;
+        let rightPos = undefined;
+        const xLocation = chartWidth - xPos + 40;
+        if (xLocation > chartWidth * 0.8) {
```

Comment:
This condition doesn't seem to be foolproof, it seems it should not depend on width ratio. If the chart is not wide, the tooltip is truncated on the left bound.

Actually, if the tooltip width is larger than half the chart width, then neither option will prevent clipping.

It seems to me the solution has to be, that if we know (with toolitip size) that it will exceed the bounds then we should snap to that edge (leftBound=0 or rightBound=0).

Reply:
The exact width of the tooltip is determined dynamically. Building a solution around calculating it before rendering might be tricky.
As the 80% threshold works well, I suggest leaving it like that for now. In case it proves not to be reliable, we attempt a different solution.